kyo-prelude/shared/src/main/scala/kyo2/Result.scala

```diff
@@ -4,134 +4,159 @@ import Result.*
                 case v                        => v
 
         // TODO avoid Option allocation
-        def unapply[A](self: Result[A]): Option[A] =
-            self.fold(_ => None)(Some(_))
+        def unapply[E, T](self: Result[E, T]): Option[T] =
+            self.fold(_ => None)(_ => None)(Some(_))
 
     end Success
 
-    case class Failure[+A](exception: Throwable)
+    sealed abstract class Cause[+E]
 
-    object Failure:
+    case class Failure[+E](error: E) extends Cause[E]
 
-        // TODO avoid Option allocation
-        def unapply[A](self: Result[A]): Option[Throwable] =
-            self.fold(Some(_))(_ => None)
+    object Failure:
+        def unapply[E, A](result: Result[E, A]): Option[E] =
+            (result: @unchecked) match
+                case result: Failure[E] => Some(result.error)
+                case _                  => None
     end Failure
 
-    extension [A](self: Result[A])
-        inline def isSuccess: Boolean =
+    case class Panic(exception: Throwable) extends Cause[Nothing]
+
+    extension [E, T](self: Result[E, T])
+
+        def isSuccess: Boolean =
             self match
-                case _: Failure[?] => false
-                case _             => true
+                case _: Cause[?] => false
+                case _           => true
 
-        inline def isFailure: Boolean = !isSuccess
+        def isFailure: Boolean =
+            self.isInstanceOf[Failure[?]]
 
-        inline def fold[B](inline ifFailure: Throwable => B)(inline ifSuccess: A => B): B =
+        def isPanic: Boolean =
+            self.isInstanceOf[Panic]
+
+        inline def fold[U](inline ifFailure: E => U)(inline ifPanic: Throwable => U)(inline ifSuccess: T => U): U =
             (self: @unchecked) match
-                case self: Failure[A] =>
-                    ifFailure(self.exception)
+                case self: Failure[E] =>
+                    ifFailure(self.error)
+                case self: Panic =>
+                    ifPanic(self.exception)
                 case _ =>
                     try
                         self match
-                            case self: SuccessFailure[A] @unchecked =>
-                                ifSuccess(self.unnest.asInstanceOf[A])
+                            case self: SuccessFailure[T] @unchecked =>
+                                ifSuccess(self.unnest.asInstanceOf[T])
                             case self =>
-                                ifSuccess(self.asInstanceOf[A])
+                                ifSuccess(self.asInstanceOf[T])
                     catch
-                        case ex if NonFatal(ex) => ifFailure(ex)
+                        case ex if NonFatal(ex) => ifPanic(ex)
 
-        inline def get: A =
-            fold(ex => throw ex)(identity)
+        def get(using E =:= Nothing): T =
+            fold(ex => throw new NoSuchElementException(s"Cause: $ex"))(throw _)(identity)
```

Comment:
Should we specialize this in the case that `E =:= Throwable`, such that the original exception is re-thrown?

Reply:
My reasoning was that `E =:= Nothing` is essentially that: if there's no typed error, `Throwable` is still present via `Panic`. A behavior similar to `Try.get`.

Reply:
https://github.com/zio/zio/blob/c7b4074d1b40727e4cca53d5b8ab3eafd6b9e549/core/shared/src/main/scala/zio/ZIO.scala#L6352-L6361

My suggestion was more so a form of specialization that requires users to be explicit with the behaviors they expect.

`getOrThrow` -> `get` - requiring `E <: Throwable`
`getOrThrowFiberFailure` -> `getOrThrowPanic` - where `E` is anything